### ISSUE TITLE
e-commerce

### ISSUE THREAD

**Issue by sc0ttj** [sc0ttj/mdsh#88]
Support creation of very simple e-commerce websites using `mdsh`.

This would include nicer/clearer ways to:

- make a product a proper "thing" with a title, description, price, payment merchant
- add payment merchant credentials to site data
- add a catalogue of products to site data (in assets/data/ folder)
- add a single product to page data (in front matter)
- to add multiple products to a page (templating)
- to add a single product to a page  (templating)

### 1. Multiple products:

a. Add in `assets/data/products.yml`:

```yml
product1
  id: my-cool-page
  name: 'Nice toaster or something' 
  price: 10
product2
  id: some-other-page
  name: 'Nice coat or something' 
  price: 20
  ```
b. To list multiple products per page:

```handlebars
{{foreach product in products}}
  Name: {{product.name}}
  ...
{{/foreach}}
```

Also see `limit`, `sort_by` and `sort_array` liquid filters for manipulating the list in the templates.

To access specific products, you'd also be able to do (probably less useful) stuff like this:

```handlebars
Name: {{lookup products[3].name}}
Price: {{lookup products[3].price | money £}}
```

c. A better way to show one product per page, create a `_product_info.mustache` template partial, containing something like:

```handlebars
{{foreach product in products | where id = $page_slug}}
  Name: {{product.name}}
  Price {{product.price | money £}}
  ...
{{/foreach}}
```

The benefit of this is that we can put only one product per page, and each page gets a different product, and we only have to define this one partial, and include it in our main templates, using something like so:

```handlebars
  {{>product_info}}
```

### 2. Adding a single product to a page

Add this in front-matter of `posts/*/*/*/my-cool-page.mdsh`:

```yaml
product
  id: my-cool-page
  name: 'Nice toaster or something' 
  price: 10  
```

And access in a `_page_product.mustache` template, like so:

```handlebars
{{#page_product_name}}
Name: {{page_product_name}}
Price: {{page_product_price | money £}}
{{/page_product_name}}
```
...which could be included in the main templates like so:

```handlebars
{{>_page_product}}
```
This is probably the best approach, as we don't need an iterator to get to the product info, or the `lookup` method, and we can iterate over product options more easily.

### 3. Available payment methods

- Buy now and one-off payments (see comments below)
- Subscriptions and recurring payments (see comments below)
- Add to cart buttons, with a simple shopping cart

#### shopping cart 

Build as a tiny `cash-js` plugin (basically a jQuery plugin), and add to `app.js`. 

Use localStorage to save items to a cart object, and a little bit of JS to detect the cart is not empty, then put a button on the page, and provide a nice popup modal cart. 

Show in cart modal popup:
-  a "Shopping cart" header
- a list of items in the cart, with a buy now button next to each one
- a "Checkout" button at the bottom

### Alternative merchants

- stripe
- snipcart

Don't tie people in to one or the other too much, or at all if possible..

**Comment by sc0ttj** [comment]
## Snipcart 

Adds a nice shopping cart, requires, JS, can define product data on your own site as HTML attributes (in our case, in the yaml, then compiled into the templates).. But very pricey - £8 a month minimum, or 4.5% commission if you sell over £500 a month.

How to add products to your pages:
- https://docs.snipcart.com/configuration/product-definition

Pricing:
- https://snipcart.com/pricing

**Comment by sc0ttj** [comment]
## PayPal 

You must generate the products on the PayPal site, to get a product id, and the generated embed code. You must then put the embed code HTML into your templates somewhere.

You can replace the hard-coded values in the embed code with with template variables, which you can define in your YAML, *but the YMAL data must map to the attributes created for that product, on the PayPal site*.

In other words,  for each product you add to `mdsh` using PayPal, you need to create a matching product on the PayPal website, and add that embed code directly each time (no templating!).

Or, you can put the embed code in your templates and replace the hard-coded values - but I've yet to test creating products via YAML only, without creating them on the PayPal site as well.

**Comment by sc0ttj** [comment]
The last 2 PayPal implementations were done and merged.. Didn't bother with Snipcart. Closing this issue.
